crypto/rand/drbg_lib.c

```diff
@@ -366,6 +376,15 @@ int RAND_DRBG_generate(RAND_DRBG *drbg, unsigned char *out, size_t outlen,
     if (drbg->reseed_counter >= drbg->reseed_interval)
         drbg->state = DRBG_RESEED;
 
+    if (drbg->parent != NULL) {
+        if (drbg->parent->lock)
+            CRYPTO_THREAD_read_lock(drbg->parent->lock);
+        if (drbg->restart_count != drbg->parent->restart_count)
+            drbg->state = DRBG_RESEED;
```

Comment:
Certainly the global public and private DRBGs should reseed when the master does.  I'm less sure about the levels below these.

There could be hundreds or thousands of DRBGs in the SSL objects and this would make a reseeding an exceedingly expensive operation.  I recognise that the child DRBGs will reseed in a lazy manner so the cost is spread over time but it's still a large cost.

Reply:
I understand your concerns about the reseeding chain reaction. As you suggested, I could enable automatic restarting only for the  two global chained DRBGs (private and public). There are three possibilities to selectively enable this auto-restart feature for DRBGs:

1.	Use `lock != 0` as an indicator that automatic restart is enabled.  =>  _Looks hacky to me_.
2.	Add an `auto_restart` member.  =>  _Adds a seldomly used member: Most DRBGs will have it disabled,  so there are two unused members, auto_restart and restart_count_.
3.	Use `restart_count != 0` as indicator that automatic restart is enabled. In order to enable auto restart, one simply sets the counter to 1.  An integer overflow check when incrementing can safely be omitted. _Before this ever overflows, madness rules_ ;-).

As usual, the third variant is my favourite choice. :-)  What is your opinion?

Reply:
4. If we aren't reseeding from a parent, reseed the global public and private DRBGs.  It's a big hacky but could be reworked to "if we're reseeding the global master, flag the other two global RNGs for reseeding".

5. Have a global variable indicating that all DRBGs should be reseeded rather than just the global ones.  It wouldn't need locking since it will essentially be read-only.  I can't think of a case where you want some but not all of the non-global DRBGs reseeded.  That usually means there are several and I'm just not thinking straight enough.